### ISSUE TITLE
Viewing logs from Troubleshooting modal 'Jumps' when running multiple instances

### ISSUE THREAD

**Issue by patricktrainer** [metabase/metabase#9846]
### Issue

Ive noticed that after another ec2 instance is added to the beanstalk environment when you view the log stream from the Admin window, the stream seems to 'jump' back and forth, almost as if there is a separate log steam for each running ec2 instance that is being mashed together in the view. Strangely though, the timestamps of the logs seem to be hours apart. 

**Expected Behavior**: One stream, independent of number of running instances

Ive attached a gif of the 'jump' back and forth as well 

### Environment

- Browser: Chrome 73.0
- OS: OSX 10.14
- Connected DB: Redshift
- Metabase version: 32.4
- Metabase hosting environment: Elastic Beanstalk
- Metabase internal database: Postgres

![mb_log](https://user-images.githubusercontent.com/36834097/56673200-74eac980-667d-11e9-9ab4-cb7271ce5813.gif)

Thanks for the help here!

**Comment by tlrobinson** [comment]
The in-app Logs feature was built before Metabase supported running multiple backend instances, and we haven't done anything to support that yet. I was wondering when someone would notice :)

The requests for new log messages are being routed to different backend instances, but we assume there's just one instance.

I see two options:

1. disable the "Logs" panel if we detect you're running multiple instances (@camsaul how can we detect that?). In that case you should probably be sending logs to a central location anyway (ELK, Papertrail, Splunk, etc).
2. do something more complicated like merging the logs on the frontend (but we can't force requests to be routed to specific instances so it might still end up being jumpy)

I think the first option is more likely what we'll want to do.

**Comment by patricktrainer** [comment]
Makes sense - glad I opened that can of worms 😄  
The in-app logs are pretty convenient, but a central spot makes sense too. 

Thanks for the reply!

**Comment by camsaul** [comment]
Pasting discussion we had on the subject here so we don't forget

##### @camsaul

The way the logger works right now is we have an extra “appender”. When logging occurs the default appender writes it to the console but our custom appender keeps the last 2500 entries in memory using a circular buffer.

When we want to show those on the FE the API endpoint just dumps out the contents of the buffer.

So instead of storing the formatted/flattened log message it would be easy enough to store the different components.

Actually right now this fn is applied to every `LoggingEvent` before stuffing it in the buffer

```clj
(defn- event->log-string [^LoggingEvent event]
  ;; for messages that include an Exception, include the string representation of it (i.e., its stacktrace)
  ;; separated by newlines
  (str/join
   "\n"
   (cons
    (let [ts    (time/unparse formatter (coerce/from-long (.getTimeStamp event)))
          level (.getLevel event)
          fqns  (.getLoggerName event)
          msg   (.getMessage event)]
      (format "%s \033[1m%s %s\033[0m :: %s" ts level fqns msg))
    (seq (.getThrowableStrRep event)))))
```

it would be next-level to just keep everything broken-out and then on the client you could filter by log level or logger name (class/namespace where logging occurred) etc

##### @tlrobinson

yeah probably makes sense for the endpoint to just return a structured object, with those 4 things plus the instance uuid
then the frontend can do whatever it wants with it, either letting you pick an instance or interleaving them

**Comment by camsaul** [comment]
The TL;DR is we'll change the way the log messages are stored and sent to the frontend client and then the frontend client can either separate out the messages by instance or interleave them and provide one combined stream

**Comment by camsaul** [comment]
@tlrobinson bonus points: we can add a "since" parameter to the API request so we don't fetch logs that you've already seen. Then instead of replacing the whole thing we can just append new messages

**Comment by tlrobinson** [comment]
We’d have to pass a mapping of UUIDs to `since` values since we don’t know which instance it will hit, but yeah

**Comment by mazameli** [comment]
Bumping this up in priority to a p1. This is making the troubleshooting logs useless for us internally when trying to use it on our Stats instance.

**Comment by camsaul** [comment]
https://github.com/metabase/metabase/wiki/Bug-Prioritization 🤔

**Comment by tlrobinson** [comment]
Yeah, I'd call this a P2.

@camsaul if you make PR with the API returning JSON objects and the instance UUID I can do the frontend.

Not sure if we should bother doing the `since` thing because as I mentioned we'd need to pass `since` values for every instance UUID.

**Comment by walterl** [comment]
Fixed in #10522
